### PR TITLE
[Do Not Merge] [SwiftSyntax] Build on Linux

### PR DESCRIPTION
**Pull request by harlanhaskins**
This is an experiment to get SwiftSyntax building on Linux. It's currently a pile o' hacks.

### CODE REVIEW

**Comment by harlanhaskins** [comment]
There has _got_ to be a better way than what I'm doing to link with Foundation on Linux.

**Comment by harlanhaskins** [comment]
The tests won't pass because they also need the same set of `-I` and `-L` flags that SwiftSyntax does.

File: tools/SwiftSyntax/CMakeLists.txt [hunk 1]
```diff
@@ -1,5 +1,29 @@
...
+
+    -L"${SWIFT_PATH_TO_LIBDISPATCH_BUILD}/src"
+    -L"${SWIFT_PATH_TO_LIBDISPATCH_BUILD}/src/swift"
+    -L"${FOUNDATION_BUILD_DIR}"
+    -L"${FOUNDATION_BUILD_DIR}/usr/lib/swift"
+    -L"${FOUNDATION_BUILD_DIR}/usr/lib/swift/CoreFoundation"
+
+    -lFoundation
+    -lCoreFoundation
+    -ldispatch)
```
Comment (compnerd) [inline tools/SwiftSyntax/CMakeLists.txt]: These three can be handled via `target_link_libraries`.

File: tools/SwiftSyntax/CMakeLists.txt [hunk 2]
```diff
@@ -1,5 +1,29 @@
...
+    -I"${SWIFT_PATH_TO_LIBDISPATCH_BUILD}/src/swift"
+    -I"${FOUNDATION_BUILD_DIR}"
+    -I"${FOUNDATION_BUILD_DIR}/usr/lib/swift"
+    -I"${FOUNDATION_BUILD_DIR}/usr/lib/swift/CoreFoundation"
+
+    -L"${SWIFT_PATH_TO_LIBDISPATCH_BUILD}/src"
+    -L"${SWIFT_PATH_TO_LIBDISPATCH_BUILD}/src/swift"
+    -L"${FOUNDATION_BUILD_DIR}"
+    -L"${FOUNDATION_BUILD_DIR}/usr/lib/swift"
+    -L"${FOUNDATION_BUILD_DIR}/usr/lib/swift/CoreFoundation"
```
Comment (compnerd) [inline tools/SwiftSyntax/CMakeLists.txt]: Please use the `swift_target_link_search_directories` function for these search paths.

File: tools/SwiftSyntax/CMakeLists.txt [hunk 3]
```diff
@@ -1,5 +1,29 @@
...
+# FIXME: This is _awful_. These pieces could/should be sunk into
+#        `add_swift_library`
+if("${SWIFT_HOST_VARIANT_SDK}" STREQUAL "LINUX")
+  LIST(APPEND swiftsyntax_compile_flags
+    -I"${SWIFT_PATH_TO_LIBDISPATCH_SOURCE}"
+    -I"${SWIFT_PATH_TO_LIBDISPATCH_BUILD}/src"
+    -I"${SWIFT_PATH_TO_LIBDISPATCH_BUILD}/src/swift"
+    -I"${FOUNDATION_BUILD_DIR}"
+    -I"${FOUNDATION_BUILD_DIR}/usr/lib/swift"
+    -I"${FOUNDATION_BUILD_DIR}/usr/lib/swift/CoreFoundation"
```
Comment (compnerd) [inline tools/SwiftSyntax/CMakeLists.txt]: These can be handled via `target_include_directories`.

Comment (harlanhaskins) [reply tools/SwiftSyntax/CMakeLists.txt]: Thanks! Is there a good way to get the full target name that `add_swift_library` creates?

Comment (harlanhaskins) [reply tools/SwiftSyntax/CMakeLists.txt]: I'm trying to thread through `INCLUDE_DIRECTORIES` in `add_swift_library`, passing it through to `target_include_directories`, and it doesn't actually adjust the command line used to compile `SwiftSyntax`.

**Pull request closed without merging**